### ISSUE TITLE
Nicer formatting for multiline strings

### ISSUE THREAD

**Issue by Nemo157** [colin-kiegel/rust-pretty-assertions#24]
While the current output is miles ahead of the default `assert_eq!` for finding the errors in multiline strings, it would be amazing if multiline strings could actually be displayed as multiline. Another extension to this would be to use raw string literals when the string contains embedded `"` characters

Current output (0.5.1):

![screen shot 2018-06-10 at 12 42 46 pm](https://user-images.githubusercontent.com/81079/41200812-0244e562-6cac-11e8-98a5-873b1a7051cf.png)

A quick sketch of what multiline and raw string literals could look like:

![screen shot 2018-06-10 at 12 51 25 pm](https://user-images.githubusercontent.com/81079/41200884-16fefae6-6cad-11e8-8a5b-f9699c517f0e.png)

I'm not sure if there are cases where this would be sub-optimal, hopefully it's possible to use a simple heuristic for detecting when to enable output like this that can be used in all cases.

**Comment by Nemo157** [comment]
Pushed a super-quick implementation of this: https://github.com/Nemo157/rust-pretty-assertions/commit/93326329cbdbc60fb54062b8c351a01aaf60c28a

Probably needs a bit better heuristics, but it at least works for my usecase:

![screen shot 2018-06-10 at 1 48 43 pm](https://user-images.githubusercontent.com/81079/41201274-0acadd50-6cb5-11e8-9040-254232b06c8f.png)

**Comment by idubrov** [comment]
Here is a non-heuristic variant we use:

```rust
/// Wrapper around string slice that makes debug output `{:?}` to print string same way as `{}`.
/// Used in different `assert*!` macros in combination with `pretty_assertions` crate to make
/// test failures to show nice diffs.
#[derive(PartialEq, Eq)]
#[doc(hidden)]
pub struct PrettyString<'a>(pub &'a str);

/// Make diff to display string as multi-line string
impl<'a> fmt::Debug for PrettyString<'a> {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    f.write_str(self.0)
  }
}
```
but you have to wrap strings manually: `assert_eq!(PrettyString("hello"), PrettyString("bye"))`.

**Comment by colin-kiegel** [comment]
@Nemo157 Yes, I would be interested in the feature - but maybe behind a feature flag (we can discuss, what the default should be). Are you still interested in contributing to this feature?

Sorry for replying so late..

**Comment by martinlindhe** [comment]
Ping @Nemo157

If you no longer wish to work on having your branch upstreamed, I am willing to help finish it up

**Comment by rfdonnelly** [comment]
I've extended @idubrov solution to automate the wrapping of strings with a `PrettyString` by adding a local version of the `assert_eq!()` macro that wraps the strings and calls `pretty_assertions::assert_eq!()`.

```Rust
/// Wrapper around string slice that makes debug output `{:?}` to print string same way as `{}`.
/// Used in different `assert*!` macros in combination with `pretty_assertions` crate to make
/// test failures to show nice diffs.
#[derive(PartialEq, Eq)]
#[doc(hidden)]
pub struct PrettyString<'a>(pub &'a str);

/// Make diff to display string as multi-line string
impl<'a> fmt::Debug for PrettyString<'a> {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    f.write_str(self.0)
  }
}

macro_rules! assert_eq {
    ($left:expr, $right:expr) => {
        pretty_assertions::assert_eq!(PrettyString($left), PrettyString($right));
    }
}
```

Now you can just call `assert_eq!()` as normal:

```Rust
assert_eq!(left, right);
```

**Comment by Herschel** [comment]
This would be a very helpful feature. I'm currently using the suggestion above by @rfdonnelly, which works great!

**Comment by lig** [comment]
@colin-kiegel it looks like adding something special to `assert_eq!` call to use diff by Display instead of diff by Debug is not a problem for users. Maybe, it would be enough to add an optional third argument to the signature, i.e. passing the method to use explicitly.

**Examples**

```rust
// uses default behavior, could include a heuristics of some kind to guess the method
assert_eq!(a, b)

// uses Debug and fails if it's impossible
assert_eq!(a, b, pretty_assertions::Method::Debug)

// uses Display and fails if it's impossible
assert_eq!(a, b, pretty_assertions::Method::Display)
```

**Comment by Wilfred** [comment]
One workaround I've found helpful is converting my strings to a vec of lines, and then using `assert_eq` with that.

**Comment by dralley** [comment]
I'm trying to compare XML and this would be a lifesaver.  It looks like there's some activity in this repo again, is this being looked at?

**Comment by colin-kiegel** [comment]
I can see there is quite some interest in a different representation of multiline string diffs and I would appreciate it _behind a feature flag or something similar_ at first - to be on the safe side.

The goal of `pretty_assertions` is to be a drop-in-replacement for `assert_eq!` with different _formatting_. This is why I try to keep it similar at its roots, IMO this implies:

* `pretty_assertions::assert_eq!` should have the same signature as the original - you should be able to switch back to the original at any time without much churn
* as the original `assert_eq!` we should compare the `fmt::Debug` output instead of `fmt::Display` - _at least by default_ - to guarantee that we are giving the same output just with different formatting

With these limitations I am open to discuss different formatting approaches.

@tommilligan What are your thoughts on this as new Co-Maintainer? :-)

**Comment by tommilligan** [comment]
I've been thinking about this in the background, as it seems to be a common theme around many issues/PRs in this repo (there are currently 3 PRs open to do with customising/tweaking the output of `pretty_assertions`). There's obviously interest in customising the output, but no two people agree on the same output format.

As I see it, the core functions of `pretty_assertions` for the user are:

- Provide macros fitting the `assert_eq` and `assert_ne` API
- Provide some level of diff/diffing algorithm
- Print a diff in a pretty manner

To this end, I think we have two options:

1. Provide tonnes of feature flags, config options etc. to make the behaviour of the library customisable
1. Refactor the existing implementation into a library of components, enabling users to build their own solutions on top of it

I would vastly prefer option 2., which I think:

- Adheres to the existing library concerns
- Is cleaner in design
- Will be more maintainable over the long term

With this design, a PR like https://github.com/colin-kiegel/rust-pretty-assertions/pull/33 could instead be written as a new crate/module, which pulls in functionality from the `pretty_assertions` library for diffing and printing, but provides it's own macro implementation.

---

In my eyes, good first steps in this design path would be:

- Pulling apart the diffing/printing logic we currently use, which is tightly linked together.
- Providing a helper to construct custom macro implementations, given a differ and printer. I have no idea if such a thing is even possible, so this would need research.

**Comment by colin-kiegel** [comment]
@tommilligan I like your line of thinking, to provide tutorials and basic building blocks for custom implementations. This sounds like a good tradeoff with high flexibility and low maintenance burden.

We can then observe whether specific extensions become particularly popular - and decide whether we want to include them later.

Alternatively to the "library"-metaphor, we could also investigate a "framework"-approach - i.e. maybe we could allow the consumer to define callback-functions and thereby overriding the default behaviour. It would be interesting to investigate the pros and cons of both approaches - both seem very promising to me. :-)

**Comment by categulario** [comment]
Hi everyone,

I was just looking for a way to display a nice diff in my asserts and found this project and was very excited. Then I looked at the default string diff and was very disappointed and eventually ran into this thread. My two cents: I bet a nice multi-line string diff is the expected default (it was for me). And while the solution mentioned by @rfdonnelly works just fine it is a problem having to first find this thread, then copy the code, then see the result you're expecting. I don't think all of the people that need this feature get this far.

That being said, If I can contribute to make it happen (behind a feature flag or something (which I would recommend be visibly documented in the first paragraphs of the readme)) I'll happily do it.

Thank you for this amazing project.

**Comment by x3ro** [comment]
First of all, I want to say thanks to everyone who has contributed to `pretty_assertions` in the past, I think it's a great library.

Since this discussion has stagnated, I just wanted to throw a proposal for a solution out there (see #88). It goes the "easy" route by introducing a new macro variant for strings, `assert_str_eq!`. This variant only works for strings, and does not escape line breaks in multi-line strings.

Of course this is not a perfect solution, because the user has to know that this additional variant exists. I think however, that it's preferable to library users having to first find out that this is expected behavior (e.g. through this issue), and then having to implement a custom solution for each project they use the crate in.

I've read @tommilligan's earlier statement of not wanting to bloat the library with lots of variants / configuration options, so I realize that this may not get merged. While I agree that a library such as this one should not try to provide everything for everyone, I think that diff'ing a plain string in a readable way is relatively basic functionality which, in my opinion, this library should provide in some way.

**Comment by tommilligan** [comment]
[`assert_str_eq`](https://docs.rs/pretty_assertions/1.1.0/pretty_assertions/macro.assert_str_eq.html) has been released in v1.1.0, which I believe should broadly cover the issues discussed here - thanks to @x3ro for implementing this and pushing it forward!

Please take a look and see if this covers your use cases. If not, feel free to open a new issue for further discussion.
